[quant][graph] Fix bug in replaceConvolutionWithConv2d

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37779 [quant][mobile] Return for conv with empty batch
* #37637 [quant][graph] Fix bug in replicateDequant
* **#37635 [quant][graph] Fix bug in replaceConvolutionWithConv2d**

Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354930](https://our.internmc.facebook.com/intern/diff/D21354930)

supriyar committed:
[quant][graph] Fix bug in replaceConvolutionWithConv2d
Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
please just make this function work for `aten::conv1d` since we already have `aten::conv1d` op

supriyar committed:
Update on "[quant][graph] Fix bug in replaceConvolutionWithConv2d"


Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
> please just make this function work for `aten::conv1d` since we already have `aten::conv1d` op

@jerryzh168 done, tested with model. We can properly test it once we support conv1d in graph mode.

supriyar committed:
Update on "[quant][graph] Fix bug in replaceConvolutionWithConv2d"


Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -34,7 +34,7 @@ c10::optional<IValue> getIValue(
 }
 
-void replaceConvolutionWithConv2d(std::shared_ptr<Graph>& graph) {
+void replaceConvolutionWithAtenConv(std::shared_ptr<Graph>& graph) {
```
@jerryzh168 @raghuramank100 From the aten op looks like the main distinguishing difference between aten::conv1d and aten::conv2d is the output_padding {0} vs {0,0} https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Convolution.cpp#L535-L543

raghuramank100 replied:
In addition, we should check for bool transposed along with output_padding to map to conv1d/2d/3d and the transpose ops.

supriyar replied:
We're already checking `transposed_value` below. When we add deconv here we can support that too.

jerryzh168 replied:
I think the length of stride/padding/dilation should be different as well, it makes more sense to check these arguments that are actually meaningful I think

raghuramank100 commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -66,17 +73,40 @@ void replaceConvolutionWithConv2d(std::shared_ptr<Graph>& graph) {
-        cudnn_enabled_value && (output_padding_value[0] == 0) &&
-        (output_padding_value[1] == 0)) {
-      return true;
+    if (output_padding_value.size() != 1) {
```
Can we also check if it is 0?

raghuramank100 reviewed (approved):
Can you add a test that checks that this code works for traced conv1d/2d/3d ?

jerryzh168 commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -66,17 +73,40 @@ void replaceConvolutionWithConv2d(std::shared_ptr<Graph>& graph) {
+    return !transposed_value && !benchmark_value && !deterministic_value &&
+        cudnn_enabled_value && (output_padding_value[0] == 0);
+  };
+  auto filter_conv2d = [](const Match& match,
+                          const std::unordered_map<std::string, Value*>& vmap) {
+    const auto& match_vmap = match.values_map;
+    auto transposed_value =
+        getIValue("transposed", match_vmap, vmap).value().toBool();
+    auto benchmark_value =
+        getIValue("benchmark", match_vmap, vmap).value().toBool();
+    auto deterministic_value =
+        getIValue("deterministic", match_vmap, vmap).value().toBool();
+    auto cudnn_enabled_value =
+        getIValue("cudnn_enabled", match_vmap, vmap).value().toBool();
+    auto output_padding_value =
+        getIValue("output_padding", match_vmap, vmap).value().toIntList();
+
+    if (output_padding_value.size() != 2) {
+      return false;
+    }
+    return !transposed_value && !benchmark_value && !deterministic_value &&
+        cudnn_enabled_value && (output_padding_value[0] == 0) &&
+        (output_padding_value[1] == 0);
   };
```
there are some duplicated code here, could you extract the common parts?

supriyar replied:
Sure

jerryzh168 reviewed (changes requested):
Please reduce code duplication

jerryzh168 commented:
yeah, probably add a replacement for conv3d as well

supriyar committed:
Update on "[quant][graph] Fix bug in replaceConvolutionWithConv2d"


Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354930](https://our.internmc.facebook.com/intern/diff/D21354930)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -51,32 +75,79 @@ void replaceConvolutionWithConv2d(std::shared_ptr<Graph>& graph) {
+                          const std::unordered_map<std::string, Value*>& vmap) {
+    std::unordered_map<std::string, c10::IValue> calc_value_map;
+    get_conv_params(match, vmap, calc_value_map);
+    if (calc_value_map["output_padding"].toIntList().size() != 1 ||
+        calc_value_map["stride"].toIntList().size() != 1 ||
+        calc_value_map["dilation"].toIntList().size() != 1) {
```
check "padding" as well, same for conv2d and conv3d.

jerryzh168 reviewed (approved):
left one inline comment. Thanks for refactoring!

jerryzh168 commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -34,7 +34,31 @@ c10::optional<IValue> getIValue(
+void get_conv_params(
+    const Match& match,
+    const std::unordered_map<std::string, Value*>& vmap,
+    std::unordered_map<std::string, c10::IValue>& calc_values) {
```
Also, why is `calc_values` passed as a reference? why not just return a new map?

jerryzh168 commented on torch/csrc/jit/passes/graph_rewrite_helper.cpp:
```diff
@@ -34,7 +34,31 @@ c10::optional<IValue> getIValue(
 }
 
-void replaceConvolutionWithConv2d(std::shared_ptr<Graph>& graph) {
+void get_conv_params(
```
also this should be getConvParams

supriyar committed:
Update on "[quant][graph] Fix bug in replaceConvolutionWithConv2d"


Summary:
replaceConvolutionWithConv2d incorrectly assumes that the size of padding is 2. For Conv1d it is 1, in which case we cannot replace with aten::conv2d

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21354930](https://our.internmc.facebook.com/intern/diff/D21354930)

[ghstack-poisoned]
